performance-tests/bench_2/node_controller/main.cpp

```diff
@@ -548,113 +557,159 @@ int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
-        return 1;
-      }
-      if (status_writer_impl->get_publication_matched_status(match) != DDS::RETCODE_OK) {
-        std::cerr << "get_publication_matched_status failed" << std::endl;
-        return 1;
-      }
-    }
-    ws->detach_condition(condition);
-
-    Status status;
-    status.node_id = this_node_id;
-    status.state = AVAILABLE;
-    status.name = name.c_str();
-    if (status_writer_impl->write(status, DDS::HANDLE_NIL)) {
-      std::cerr << "Write status failed" << std::endl;
-      return 1;
-    }
+  Status status;
+  status.node_id = this_node_id;
+  status.state = state;
+  status.name = name.c_str();
+  if (status_writer_impl->write(status, DDS::HANDLE_NIL)) {
+    return false;
   }
+  return true;
+}
 
-  WorkerManager worker_manager(process_manager);
+bool wait_for_scenario_data(AllocatedScenarioDataReader_var config_reader_impl) {
 
-  // Wait for Our Worker Configs
-  bool waiting = true;
-  while (waiting) {
-    DDS::ReturnCode_t rc;
+  DDS::ReadCondition_var read_condition = config_reader_impl->create_readcondition(
+    DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ALIVE_INSTANCE_STATE);
+  DDS::WaitSet_var ws(new DDS::WaitSet);
+  ws->attach_condition(read_condition);
 
-    DDS::ReadCondition_var read_condition = config_reader_impl->create_readcondition(
-        DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ALIVE_INSTANCE_STATE);
-    DDS::WaitSet_var ws = new DDS::WaitSet;
-    ws->attach_condition(read_condition);
+  while (!read_condition->get_trigger_value()) {
     DDS::ConditionSeq active;
-    rc = ws->wait(active, {DDS::DURATION_INFINITE_SEC, DDS::DURATION_INFINITE_NSEC});
-    ws->detach_condition(read_condition);
-    config_reader_impl->delete_readcondition(read_condition);
-    if (rc != DDS::RETCODE_OK) {
+    const DDS::Duration_t wake_interval = { 0, 500000000 };
+    DDS::ReturnCode_t rc = ws->wait(active, wake_interval);
+    if (rc != DDS::RETCODE_OK && rc != DDS::RETCODE_TIMEOUT) {
       std::cerr << "Wait for node config failed" << std::endl;
-      return 1;
+      return false;
+    }
+  }
+
+  ws->detach_condition(read_condition);
+  config_reader_impl->delete_readcondition(read_condition);
+
+  return true;
+}
+
+void wait_for_full_scenario(
+  const std::string& name,
+  NodeId this_node_id,
+  StatusDataWriter_var status_writer_impl,
+  AllocatedScenarioDataReader_var config_reader_impl,
+  Bench::TestController::AllocatedScenario& result)
+{
+  using Builder::ZERO;
+
+  Bench::TestController::AllocatedScenario allocated_scenario;
+  allocated_scenario.scenario_id = TAO::String_Manager();
+  allocated_scenario.launch_time = ZERO;
+  std::chrono::system_clock::time_point initial_attempt;
+
+  bool complete = false;
+  while (!complete) {
+    while (!wait_for_scenario_data(config_reader_impl)) {
+      // There was an actual DDS failure of some kind (not just a timeout), give it a few seconds and retry
+      std::this_thread::sleep_for(std::chrono::seconds(3));
```

Comment:
It waits 3 seconds and tries again until the process is killed. But DDS failures should be pretty rare. Most of the time, it's in WaitSet::wait() waiting for the next scenario data. If 30 seconds expires between "choosing" the scenario id and receiving a launch time, the next scenario data that comes in will reset the chosen scenario id and it will start the wait for a launch time over again. It's not the cleanest solution, but I was trying to avoid changing too much of the IDL for fear of unintended consequences.